---
title: "Quantitative Genetics in R"
output: html_document
---

## Introduction

For this module we will be using R and Rstudio. The lesson assumes no prior knowledge of R, but some experience may be helpful to some learners. At the end of the lesson, learners will be able to:

-Open and type commands into R and/or Rstudio, and have them executed in the console

-Know how to get help with R errors or warnings

-Have an understanding that there are user/community contributions called 'packages' that can help with specific areas

-Calculate the phenotypic covariance matrix

-Compute the additive co-ancestry matrix a set of markers

-Impute missing marker data

-Estimate population structure using a PCA

-Run a GWAS for a trait of interest


# Setup (to be done prior to module)

For this module you will need to download base R and rstudio (if you don't already have them). For this lesson we will be using R 3.4.3 "Kite-Eating Tree". There are distributions of base R for all operating major operating systems.

# **Download R first (to be done prior to module)**

-MAC OS users go here: https://cran.r-project.org/bin/macosx/R-3.4.3.pkg

-Windows users go here: https://cran.r-project.org/bin/windows/base/


# **Download RStudio (to be done prior to module)**
  
Rstudio is an IDE (integrated development environment) for R, and it streamlines writing of scripts, project management, and plotting - click the link below and choose "Rstudio desktop" (free version) and choose your OS version:

https://www.rstudio.com/products/rstudio/download/#download


# **Download Data Set (to be done prior to module)**

For this module we will be using a Picea glauca (white spruce) dataset of 15 phenotypes, 1694 individual trees, and some 6000 SNP markers. The markers have no linkage map, nor does the data set include pedigree based information. Picea glauca like most conifers, has a very large estimated genome size of 16.15 pg or 2100cM (estimated genetic map size), and 12 autosomal chromosomes. This means that we have very low marker coverage - this is further detailed in the paper (Beaulieu et al. 2014). White spruce is widespread in the Boreal and used mostly in residential home construction in Canada and paper making. 

We will use a slightly altered version of the original dataset for the sake of time; however, the original dataset and associated paper by Beaulieu et al. (2014) lives here if anyone is interested in taking a look.

http://datadryad.org/resource/doi:10.5061/dryad.6rd6f


# Part 1: Getting used to and oriented to R

### **Open up RStudio**
Since this is the first time we've opened R let's list the objects in our
workspace:
```{r}
ls()
```

There should be no objects in your workspace. If there are you can wipe the workspace in rstudio using the tiny *broom* icon in the environment panel. Or, you can simply type this command (**NOTE only use this command if you are certain you want to wipe all objects from the workspace**). 

```{r}
rm(list=ls())

ls()
```

You can type these directly in the console, or preferrably in our script file. If we type commands in our console, they will be more difficult to recall or remember.

## Load the phenotype data and understand its structure

Now that are workspace is a blank space we can start working with our spruce dataset. First we will load the phenotypes into the workspace:

```{r}
phenos <- read.table("S1.txt", header = TRUE, sep="\t")
```

If you're new to R, you'll notice that when you executed this command on the console - nothing happened. That's generally good news, as it means your command was successfully executed. You'll notice that if you type ```ls()``` an object named ```"phenos"```. 

The first thing to do before examining any data is examine it and visualize it. R has several useful commands to do this. First, let's figure out the size of our data. 

```{r}
dim(phenos)
```

The output of this command tells us that we have an array of 1694 rows x 15 columns. The rows always come first, columns second in R. That's great; however, we would like to get a better picture of the data. We'll do that with the command ```str()``` or the structure command.

```{r}
str(phenos)
```

This not only gives us the dimensions of the data, the datatype we are dealing with, the column names, and the column classes. What does ```str()``` mean though? We can ask R. Type the command below for any function and you'll get the help file to pop up with any arguments that might be contained within that fuction.

```{r}
?str()
```

There's also another function ```summary()``` which produces a brief summary of all data within our ```data.frame()``` columns.

```{r}
summary(phenos)
```

The ```summary()``` can be very useful for figuring out if there's a difference between mean and median values for example. 

## Phenotypic correlations and phenotypic covariance matrix

Since phenotypic correlations are important to consider in multivariate breeders' selection we can produce this matrix in R.

```{r, eval=FALSE}
cor(phenos)
```

An error is generated ```Error in cor(phenos) : x must be numeric```. If you remember from the ```str``` and ```summary``` commands, one of the columns is not numeric.

If we re-execute our command, this time without that column, it should work. We can do this with indexing. 

```{r}
cor(phenos[,-1])
```

We can exclude multiple columns by using the combine function ```c()```

```{r}
cor(phenos[,-c(1:3)])
cor(phenos[,-c(1,4:9,10)])
```

Let's generate the phenotypic covariance matrix, that we could actually use in a breeding program.

```{r}
pheno_covar <- cov(phenos[,-c(1:3)])
pheno_cor <- cor(phenos[,-c(1:3)])
```

## Visualize the data

R comes with numerous helpful commands for plotting and visualizing data. For our phenotypic dataset, let's use a couple basic plots to get a sense of our trait data

```{r}
hist(phenos$cryst)
plot(phenos$coars ~ phenos$cryst)
pairs(pheno_cor)
```

## Using packages 
One of the other positives with using R is that its open-source community contributes many pacakges. A package is a set of functions contributed by a/many users for a specific area of study. Packages may have dependencies on other packages. R has a very active community, and there are many thousands of packages that users can take advantage of, the field of genetics is no exception.

For this lesson, we will use a couple of pkgs ```rrBLUP``` from Jeff Endelman's lab at the University of Wisconsin Madison. We will install the package and load it into the R workspace with the function:

```{r, eval = FALSE}
install.packages("rrBLUP")
library(rrBLUP)
```

If we ever want to read more about a package we can use the ```citation``` function, that is: ```citation(rrBLUP)``` to read more about who generated the package or ```help()```, that is: ```help(rrBLUP)```.

```{r}
citation("rrBLUP")
help(rrBLUP)
```


# Part 2: Quantitative genetics in R

## Load the genotype data

R can read in different data types, and R packages expand that capability. Above, we read in a tab-delimited text file and it's not uncommon to read in csv (comma separated value files). There are even numerous packages that allow you to read in native '.xls' and/or '.xlsx' files. 

We will load an '.RData' file, which has been pre-processed in advance, which contains a single object ```genos```, however note that RData files can contain multiple objects if you choose. We'll also load the linkage map (this is a fake made up linkage map, with a 13th chromosome for one of the organelles or unknown SNPs).

```{r}
load("genos.RData")
dim(genos.rrBLUP)
marker.map <- read.csv("map.csv", header = TRUE)
marker.map <- marker.map[,-1]
```

What happens if we use the functions above to get an idea of the data structure? Type them in and see what happens.

Could we use our indexing technique from the phenotypic correlations to get an idea? 

On your own see if you can print a small subset of the object to get an idea of its structure.

## Compute the additive matrix and impute markers

We have no information on pedigree with this dataset, and as such we really can't differentiate between IBD (identity-by-descent) and IBS (identity-by-state). The more SNPs, the more likely IBS starts to approximates IBD. Nonetheless, we'll proceed with this toy example. We'll use the function ```A.mat``` from the ```rrBLUP``` package we previously loaded. Let's first look at what this function does by typing its name in the console, and we can also find out more by doing what?

```{r}
library(rrBLUP)
A.mat.spruce <- A.mat(genos.rrBLUP, min.MAF=0.05,
                      impute.method="mean", 
                      return.imputed=TRUE)
dim(A.mat.spruce)
```

You'll notice here that I asked also for the option to return an imputed matrix based on the mean of each marker column. This is for downstream analyses, as many GS (genomic selection) and GWAS canned packages/functions require no missing data. I also chose the method that just happens to be the quickest for imputation - NOT THE MOST ACCURATE.

What type of object is retured? It's a list, which is a special data type that allows the storing of different object types, of different lengths. 

We can ```unlist()``` both of those objects for different purposes. For the next step we'll want to use the imputed marker matrix.

```{r}
marker_imputed <- unlist(A.mat.spruce$imputed)
```

## Get an estimate of population structure

We don't know much about this dataset, save that it's a bunch of open pollinated individual trees in Canada, whose parents originated from three different latitudes/altitudes in the Province of Quebec and are included in the data set. 

While kinship is probably the most important type of structure to consider it's useful to get an estimate of population structure nonetheless. We will use principal component analysis (PCA), not to be confused with principal co-ordinate analysis (PCoA). Here we are using one of the base functions for principal co-ordinate analysis, but for larger dataset there are other options: EIGENSTRAT, flash-pca, SNPRelate (another R package), and many more.

We will use another nice plotting library to look at our PCs - ggplot2. What do you notice about the plot that strikes you given the limited information you have on the lines?


```{r, eval=FALSE}
pca.spruce <- prcomp(marker_imputed)
str(pca.spruce)

library(ggplot2)
PCi<-data.frame(pca.spruce$x)
ggplot(PCi,aes(x=PC1,y=PC2)) +
   geom_point(size=3,alpha=0.5)

ggplot(PCi,aes(x=PC1,y=PC2)) +
   geom_point(size=3,alpha=0.5, color="orange") + theme_classic()
```


## Perform a GWAS 
Even though we have a pretty good estimate of coancestry and population structure, and it's normally a good idea to include these in GWAS, here we will perform a GWAS with kinship included as the additive co-ancestry GWAS due to the local computation limits of this exercise. If you would like, you can try including the coancestry matrix (I have) and the PCs in your own GWAS. We'll choose the tree height from 1997, at 22 years of age, and include the tree id in a ```data.frame```.

```{r}
?GWAS
pheno <- read.table("S1.txt", header=TRUE)
h97 <- data.frame(pheno$Tree, pheno$h97)
markerT <- data.frame(t(marker_imputed))
colnames(markerT) <- h97$pheno.Tree
genogwas <- marker.map[marker.map$Marker_Name %in% rownames(markerT),]
genogwas <- data.frame(genogwas, markerT)
```

Ok, now let's run our GWAS, we'll remove the default min.MAF = 0.05, since we already did this when we calculated the additive matrix.

```{r, eval=FALSE}
GWAS(pheno = h97, geno = genogwas, 
                   min.MAF = 0.05, 
                   K = A.mat.spruce$A,
                   n.core = 4,
                   P3D = TRUE,
                   plot = TRUE)

```





# Lesson "Easter Eggs" (not actually covered in the lesson) 

Sometimes it's useful to gather scripts that you can use or modify for the task at hand. A few of these are provided below, but we won't cover them today. You may want to try it out at a later date.

## Recode nucleotide data to a numeric format

Recoding SNP nuleotide data is one of the things that is a common requisite for many packages for quantitative and population genetics. E.g. taking "AA", "CC", "AC", and translating it to 2, 1, 0 or 1,0,-1. There are other formats too. The function below can do this quickly (try it out later if you'd like).

The original spruce dataset from Data Dryad is loaded into the workspace, we define our function, and then we apply it to the ```genos```.

```{r, eval=FALSE}
genos <- read.table("S2.txt", header = TRUE, sep="\t", stringsAsFactors = TRUE)

genos[1:5,1:5]

recode.SNP.rrBLUP <- function(x) {
  alleles <- unique(x)
  y <- rep(0,length(x))
  y[which(x==alleles[1])] <- -1
  y[which(x==alleles[2])] <- 0
  y[which(x==alleles[3])] <- 1
  y[which(x=="N")] <- NA
  return(y)
}


X <- apply(genos[,-1],2,recode.SNP.rrBLUP)

X[1:5,1:5]
```

If you want the 0, 1, 2 format that is often the required by other programs, you can simply:

```{r, eval = FALSE}
Xadd <- X + 1

Xadd[1:5,1:5]
```


## GWAS by brute force - with help from R packages

Caution: only use this if you're fairly certain you have little to no population structure or shared ancestry. It's fast, and it's also useful because it shows you the simple linear model mechanics of GWAS! 

